src/transformers/models/mamba2/modeling_mamba2.py

```diff
@@ -0,0 +1,762 @@
+        discrete_B =  dB * hidden_states[:,:,:,:,None]
+        # 3.c perform the recurrence y ← SSM(A, B, C)(x)
+        scan_outputs = []
+        for i in range(seq_len):
+            ssm_state = ssm_state * dA[:,i,:,:] + discrete_B[:, i, :, :] # (batch, dim, dstate
+            scan_output = torch.matmul(ssm_state.to(C.dtype) , C[:,i,:].unsqueeze(-1))
+            scan_outputs.append(scan_output[:,:,:,0])
+        scan_output = torch.stack(scan_outputs, dim=1)                                # [batch, intermediate_size, seq_len]
+        scan_output = scan_output + (hidden_states * self.D[:,None])
+        scan_output = self.norm(scan_output.view(batch_size, seq_len, -1), gate)
+        scan_output = torch.cat([nn.functional.silu(z0) * x0, scan_output], dim=-1)
+        if cache_params is not None:
+            cache_params.ssm_states[self.layer_idx].copy_(ssm_state)
+
+        # 4. Final linear projection
+        contextualized_states = self.out_proj(scan_output.to(dtype))  # [batch, seq_len, hidden_size]
+        return contextualized_states
+    # fmt: on
+
+    def forward(self, hidden_states, cache_params: Optional[Mamba2Cache] = None):
+        if is_fast_path_available:
+            return self.cuda_kernels_forward(hidden_states, cache_params)
+        return self.slow_forward(hidden_states, cache_params)
+
+
+# Copied from transformers.models.mamba.modeling_mamba.MambaRMSNorm with Mamba->Mamba2
+class Mamba2RMSNorm(nn.Module):
+    def __init__(self, hidden_size, eps=1e-6):
+        """
+        Mamba2RMSNorm is equivalent to T5LayerNorm and LlamaRMSNorm
+        """
+        super().__init__()
+        self.weight = nn.Parameter(torch.ones(hidden_size))
+        self.variance_epsilon = eps
+
+    def forward(self, hidden_states):
+        input_dtype = hidden_states.dtype
+        hidden_states = hidden_states.to(torch.float32)
+        variance = hidden_states.pow(2).mean(-1, keepdim=True)
+        hidden_states = hidden_states * torch.rsqrt(variance + self.variance_epsilon)
+        return self.weight * hidden_states.to(input_dtype)
+
+
+# Copied from transformers.models.mamba.modeling_mamba.MambaBlock with Mamba->Mamba2
+class Mamba2Block(nn.Module):
+    def __init__(self, config, layer_idx):
+        super().__init__()
+        self.config = config
+        self.layer_idx = layer_idx
+        self.residual_in_fp32 = config.residual_in_fp32
+        self.norm = Mamba2RMSNorm(config.hidden_size, eps=config.layer_norm_epsilon)
+        self.mixer = Mamba2Mixer(config, layer_idx=layer_idx)
+
+    def forward(self, hidden_states, cache_params: Optional[Mamba2Cache] = None):
+        residual = hidden_states
+        hidden_states = self.norm(hidden_states.to(dtype=self.norm.weight.dtype))
+        if self.residual_in_fp32:
+            residual = residual.to(torch.float32)
+
+        hidden_states = self.mixer(hidden_states, cache_params=cache_params)
+        hidden_states = residual + hidden_states
```

Comment:
I think I've had it wrong here since the norm takes care of the downcast... mb

It will lead to fp32 in the last output of the block though which might be avoided by a downcast. For example, a bf16 input is computed correctly but the last output given to the rms norm ("norm_f") will take in fp32 and subsequently gets fp32 again. This leads to dtype mismatches if we expect a bf16 output for other subsequent components. 

In short, calculations for fp16/bf16 are correct but we get fp32 as the final output which shouldn't be the case imo.

Reply:
cc @molbap not sure about this one but would love to hear your opinion on this.